Firestore/Source/Public/FirebaseFirestore/FIRTransaction.h

```diff
@@ -29,8 +29,8 @@ NS_ASSUME_NONNULL_BEGIN
 NS_SWIFT_NAME(Transaction)
 @interface FIRTransaction : NSObject
 
-/** :nodoc: */
-- (id)init __attribute__((unavailable("FIRTransaction cannot be created directly.")));
+/** Initialiser. Only for use in unit tests. */
+- (id)init;
```

Comment:
Types like this look a bit different from the simple data types like `FIRTimestamp` or `FIRInAppMessagingDisplayMessage` where exposing the designated initializers looks natural and useful. In this case exposing the init actually may lead to confusion and reduce quality of the API. 

I think for types like this we should take a step forward and provide a fake implementation. We can have a separate module (e.g. `FirebaseTestsSupport`, or multiple like `FirebaseFirestoreTestSupport`, `FirebaseInAppMessagingTestsSupport`, etc.) with Objective-C implementations of the fakes which would subclass the original types and encapsulate the dirty work. In this case this module/podspec can be added as a dependency to tests only, so we can keep the public API clean but unblock testing.

@peterfriese @ryanwilson @paulb777 @ncooke3 WDYT?

Reply:
SGTM

Reply:
this approach SGTM as well!

Reply:
SG - TBH, I opened up every class that looked remotely interesting, in an attempt to make sure @lorenzofiamingo is unblocked. I realise this might be overly broad, so I might lock the simple types down again and just keep the others open.

Reply:
SGTM to add `init` methods to unblock contribution for now with TODO to refactor it to be available only in a separate testing support target only.